include/aws/common/array_list.inl

```diff
@@ -147,6 +147,7 @@ int aws_array_list_front(const struct aws_array_list *AWS_RESTRICT list, void *v
         "Input pointer [val] must point writable memory of [list->item_size] bytes.");
     if (aws_array_list_length(list) > 0) {
         memcpy(val, list->data, list->item_size);
+        AWS_POSTCONDITION(AWS_BYTES_EQ(val, list->data, list->item_size));
```

Comment:
Is this really necessary? This seems like it would add O(n) cost to each call to this function in debug builds. Are we really concerned that memcpy could fail?

Reply:
Thanks for the comment! 

The reason why I added this, is to extend the function contract (what we know about a function by just reading its pre and postconditions). It is true that the `memcmp` would have to be executed on debug builds, but we are already doing a `memcpy` so, it shouldn't be a big issue. 

If the cost of a `memcmp` there is prohibitive in debug builds, then I could wrap it in `AWS_DEEP_CHECKS` macro, to only be enabled then.

In any case, I have made a PR in CBMC (https://github.com/diffblue/cbmc/pull/4976) that will allow us to remove all `AWS_POSTCONDITION`s from the code, and just have them as contracts -- which means that they will not be executed in debug builds.

Reply:
I would prefer it go in a DEEP_CHECKS check if you don't mind.

Reply:
I wrapped it in a deep check.